### ISSUE TITLE
Cannot accurately set BrowserWindow position / size when DPI scaling is not 100%

### ISSUE THREAD

**Issue by dimdimdimdim** [electron/electron#10659]
**Background**

I am trying to make a popup window (child frameless BrowserWindow) appear at a specific position related to a DOM element of another window. I managed to get the child window located [almost] appropriately using getContentBounds() on the main window, adding the DOM element position given from element.getBoundingClientRect(), and setting the window position using setPosition().

**What goes wrong**

When the DPI scaling factor is not 100%, on Windows (for example 125%, configured in the display settings), the child window position is off by a few pixels. Obviously, this is because the window position must be given in CSS pixels units (and not actual physical pixels), and the getContentBounds / setPosition / etc... methods all return and take integral numbers as parameters. The window position is therefore rounded arbitrarily and we loose accuracy.

**Solution ?**
A possible solution could possibly be for the getContentBounds / setPosition /etc... methods to allow the use of floating point numbers, and only round them *after* translating them to physical pixel units.

Note: this would also probably solve this problem: https://github.com/electron/electron/issues/9477

**Comment by dimdimdimdim** [comment]
Just some additional notes, for the people interested: I found a possible workaround, with an unsolved sub-problem:

If I force the scale factor to 1 from the start of the main process using:

    app.commandLine.appendSwitch('high-dpi-support', 'true');
    app.commandLine.appendSwitch('force-device-scale-factor', '1');

I can, oviously, avoid the problem. But the window content is much smaller. So I use

    webPreferences: { zoomFactor: 1.25 } }

when creating my windows to work around this, and adjust the popup positioning from the main process by doing something like:

	let r = mainWindow.getContentBounds();
	popupWindow.setPosition(Math.round(elementPos.x * 1.25 + r.x), Math.round(elementPos.y * 1.25 + r.y));

Great! Everything is fine. But the 1.25 scale factor is now hardcoded. I could obtain it by using:

    require('electron').screen.getPrimaryDisplay().scaleFactor;

But this returns the forced scale factor (1) if you do it from the app ready event. And accessing screen isn't allowed before the app is ready. And the "app.commandLine.appendSwitch" calls are, however, not allowed after the app is ready.

Dead end.

**Comment by codebytere** [comment]
heyo, in order for us to be able to look at this we need to know, in addition to what you described above:
**your electron version**
**your current OS & its version**

**Comment by dimdimdimdim** [comment]
@codebytere Sorry. Here you are:
Electron version v1.7.5
Windows 10

But I looked at the electron and chromium cource code, it seems that having floats instead of integers for the window positioning/sizing API will be a lot of work. Chromium itself has the widget coordinates defined as integers everywhere. By the way, from what I've seen, it probably doesn't depend on a specific electron version or OS version (except maybe MacOS).

Anyway, I managed to solve the final step in my previous workaround: I am not setting the "--force-device-scale-factor" command line option, and I can therefore get the scaleFactor right. Then, if the scale factor isn't 1, I simply restart the whole application using app.relaunch, effectively adding "--force-device-scale-factor=1" to the relaunched process, and also passing the original scale factor as an additional parameter. I can then go on. But, obviously, it slows down app launching.

There has been one more problem: due to this other bug: https://github.com/electron/electron/issues/10572, the zoomFactor setting passed when creating the BrowserWindow is made persistent, and ignored on subsequent app launches. So if the system scale factor settings changed, I got stuck at the previous zoomFactor even when relaunching the app. The solution for this specific part of the problem was to set the zoomFactor from the renderer process using webFrame.setZoomFactor.

Yeah, this is a very convoluted workaround, overall.

Finally, if somebody ever wants to go this route, one must know that this leads to the native menu being much smaller, and the DevTools too. In fact, anything that is not the browser part of the window gets a scale factor of 1. This isn't a problem for me, as the whole point of displaying these popup windows is actually to replace the menu with something more flexible.
